ctapipe/io/hdf5merger.py

```diff
@@ -0,0 +1,467 @@
+                )
+            except Exception:
+                raise CannotMerge(
+                    f"CTA Rerence meta not found in existing output file: {self.output_path}"
+                )
+
+            self.data_model_version = self.meta.product.data_model_version
+
+            # focal length choice doesn't matter here, set to equivalent so we don't get
+            # an error if only the effective focal length is available in the file
+            self.subarray = SubarrayDescription.from_hdf(
+                self.h5file, focal_length_choice=FocalLengthKind.EQUIVALENT
+            )
+            self.required_nodes = _get_required_nodes(self.h5file)
+
+    def __call__(self, other: Union[str, Path, tables.File]):
+        """
+        Append file ``other`` to the ouput file
+        """
+        exit_stack = ExitStack()
+        if not isinstance(other, tables.File):
+            other = exit_stack.enter_context(tables.open_file(other, mode="r"))
+
+        with exit_stack:
+            self._check_can_merge(other)
+
+            Provenance().add_input_file(other.filename)
+            try:
+                self._append(other)
+            finally:
+                self._update_meta()
+
+    def _update_meta(self):
+        # update creation date and id
+        time = Time.now()
+        id_ = str(uuid.uuid4())
+        self.meta.product.id_ = id_
+        self.meta.product.creation_time = time
+
+        with warnings.catch_warnings():
+            warnings.simplefilter("ignore", tables.NaturalNameWarning)
+            self.h5file.root._v_attrs["CTA PRODUCT CREATION TIME"] = time.iso
+            self.h5file.root._v_attrs["CTA PRODUCT ID"] = id_
+        self.h5file.flush()
+
+    def _check_can_merge(self, other):
+        try:
+            other_meta = metadata.Reference.from_dict(metadata.read_metadata(other))
+        except Exception:
+            raise CannotMerge(
+                f"CTA Rerence meta not found in input file: {other.filename}"
+            )
+
+        if self.meta is None:
+            self.meta = other_meta
+            self.data_model_version = self.meta.product.data_model_version
+            metadata.write_to_hdf5(self.meta.to_dict(), self.h5file)
+        else:
+            other_version = other_meta.product.data_model_version
+            if self.data_model_version != other_version:
+                raise CannotMerge(
+                    f"Input file {other.filename:!r} has different data model version:"
+                    f" {other_version}, expected {self.data_model_version}"
+                )
+
+        for node_path in self.required_nodes:
+            if node_path not in other.root:
+                raise CannotMerge(
+                    f"Required node {node_path} not found in {other.filename}"
+                )
+
+    def _append(self, other):
+        # Configuration
+        self._append_subarray(other)
+
+        key = "/configuration/observation/scheduling_block"
```

Comment:
This could be a loop over keys.
There are very few special cases here

Reply:
Every case is a special case in which options are looked at to decide what is included or not...

Reply:
I am open for another solution, but I tried and it resulted in the equivalent number of lines in a dict that had keys, categories, and what is stored (table, algorithm group, tel table group, etc.)

Reply:
But is this not 
```python
for key in keys:
    if key in other.root:
        self.required_nodes.add(key)
        self._append_table(other, other.root[key])
```
for most of them?
And then you could still keep the special cases as they are right now

Reply:
Not, you are missing the traitlets that select if that particular key should be written

Reply:
and the different kinds of `_append_XXX`

Reply:
I think at least a few could be collapsed into loop within each option group, but of course that doesn't do so much to simplify the code. You could make a dict of dicts to map the options to the keys to check related to that option, but again, it may be less readable that way, so maybe not worth the effort.

Reply:
I actually had that and found it less readable, but I am open to suggestions...

Reply:
I merged the ones which are exactly the same, didn't reduce it by much though.